Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift

```diff
@@ -356,4 +357,100 @@ class HTTPServerPipelineHandlerTest: XCTestCase {
+                        self.state = .req1EndExpected
+                        sendEnd = false
+                    case "/req_2":
+                        XCTAssertEqual(.req2HeadExpected, self.state)
+                        self.state = .req2EndExpected
+                    case "/req_3":
+                        XCTAssertEqual(.req3HeadExpected, self.state)
+                        self.state = .req3EndExpected
+                    case "/req_boom":
+                        XCTAssertEqual(.reqBoomHeadExpected, self.state)
+                        self.state = .reqBoomEndExpected
+                    default:
+                        XCTFail("didn't expect \(head)")
+                    }
+                    ctx.write(self.wrapOutboundOut(.head(HTTPResponseHead(version: .init(major: 1, minor: 1), status: .ok))), promise: nil)
+                    if sendEnd {
+                        ctx.write(self.wrapOutboundOut(.end(nil)), promise: nil)
+                    }
+                    ctx.flush()
+                case .end:
+                    switch self.state {
+                    case .req1EndExpected:
+                        self.state = .req2HeadExpected
+                    case .req2EndExpected:
+                        self.state = .req3HeadExpected
+
+                        // this will cause `channelRead` to be recursively called and we need to make sure everything then still works
+                        try! (ctx.channel as! EmbeddedChannel).writeInbound(HTTPServerRequestPart.head(HTTPRequestHead(version: .init(major: 1, minor: 1), method: .GET, uri: "/req_boom")))
+                        try! (ctx.channel as! EmbeddedChannel).writeInbound(HTTPServerRequestPart.end(nil))
+                    case .req3EndExpected:
+                        self.state = .reqBoomHeadExpected
+                    case .reqBoomEndExpected:
+                        self.state = .done
+                    default:
+                        XCTFail("illegal state for end: \(self.state)")
+                    }
+                case .body:
+                    XCTFail("we don't send any bodies")
+                }
+            }
+        }
+
+        let handler = VerifyOrderHandler()
+        XCTAssertNoThrow(try channel.pipeline.add(handler: handler).wait())
+
+        for f in 1...3 {
+            XCTAssertNoThrow(try self.channel.writeInbound(HTTPServerRequestPart.head(makeRequestHead(uri: "/req_\(f)"))))
+            XCTAssertNoThrow(try self.channel.writeInbound(HTTPServerRequestPart.end(nil)))
+        }
+
+        // now we should have delivered the first request, with the second and third buffered
+        XCTAssertEqual(.req2HeadExpected, handler.state)
+
+        // finish 1st request, that will send through the 2nd one which will then write the 'req_boom' request
+        XCTAssertNoThrow(try channel.writeAndFlush(HTTPServerResponsePart.end(nil)).wait())
+
+        XCTAssertEqual(.done, handler.state)
```

Comment:
Is this end case right? It doesn't look right, why are we expecting to see all four requests when only one `end` has been sent through?

Reply:
all of them are going through, only `req_1` is delayed